From 7b2b0dfa3dc30b6afce7dae875dcacd619541210 Mon Sep 17 00:00:00 2001
From: Tobias <5702338+T0biii@users.noreply.github.com>
Date: Sat, 15 Mar 2025 22:06:43 +0000
Subject: [PATCH] [PATCH] rockchip: rk3399 disable PHY_LANE_IDLE_OFF for each
 instance  of rockchip_pcie_phy_power_one

This patch fixes an issue in the Rockchip PCIe PHY driver where, after a warm restart of the rockchip_pcie_phy module, PCIe lanes other than lane 0 could remain stuck in the PHY_LANE_IDLE_OFF state. This resulted in the PCIe link being restricted to x1 mode, even in configurations designed to use multiple lanes. Patch forked from Armbian (for further details see armbian/build#7572).

Signed-off-by: Juergen Wurzer <wurzer.juergen@gmail.com>
---
 ...disable-PHY_LANE_IDLE_OFF-for-each-i.patch | 65 +++++++++++++++++++
 1 file changed, 65 insertions(+)
 create mode 100644 patches/openwrt/0011-rockchip-rk3399-disable-PHY_LANE_IDLE_OFF-for-each-i.patch

diff --git a/patches/openwrt/0011-rockchip-rk3399-disable-PHY_LANE_IDLE_OFF-for-each-i.patch b/patches/openwrt/0011-rockchip-rk3399-disable-PHY_LANE_IDLE_OFF-for-each-i.patch
new file mode 100644
index 00000000..1d55bee8
--- /dev/null
+++ b/patches/openwrt/0011-rockchip-rk3399-disable-PHY_LANE_IDLE_OFF-for-each-i.patch
@@ -0,0 +1,65 @@
+From 9abc27b05903b081e6cc35eb2571fdd3a41e3f4f Mon Sep 17 00:00:00 2001
+From: Juergen Wurzer <wurzer.juergen@gmail.com>
+Date: Sat, 15 Mar 2025 22:39:14 +0100
+Subject: [PATCH] rockchip: rk3399 disable PHY_LANE_IDLE_OFF for each instance
+ of rockchip_pcie_phy_power_one
+
+This patch fixes an issue in the Rockchip PCIe PHY driver where, after a warm restart of the rockchip_pcie_phy module, PCIe lanes other than lane 0 could remain stuck in the PHY_LANE_IDLE_OFF state. This resulted in the PCIe link being restricted to x1 mode, even in configurations designed to use multiple lanes. Patch forked from Armbian (for further details see armbian/build#7572).
+
+Signed-off-by: Juergen Wurzer <wurzer.juergen@gmail.com>
+---
+ .../402-2-rk3399-fix-pci-lanes.patch          | 42 +++++++++++++++++++
+ 1 file changed, 42 insertions(+)
+ create mode 100644 target/linux/rockchip/patches-6.6/402-2-rk3399-fix-pci-lanes.patch
+
+diff --git a/target/linux/rockchip/patches-6.6/402-2-rk3399-fix-pci-lanes.patch b/target/linux/rockchip/patches-6.6/402-2-rk3399-fix-pci-lanes.patch
+new file mode 100644
+index 0000000000..6fed47bc8f
+--- /dev/null
++++ b/target/linux/rockchip/patches-6.6/402-2-rk3399-fix-pci-lanes.patch
+@@ -0,0 +1,42 @@
++From 77f6dfcb20c2dc6a4a2f5303709c6fa0c7b65f30 Mon Sep 17 00:00:00 2001
++From: Valmantas Paliksa <walmis@gmail.com>
++Date: Thu, 12 Dec 2024 12:24:33 +0200
++Subject: [PATCH] Disable PHY_LANE_IDLE_OFF for each instance of rockchip_pcie_phy_power_one
++
++This patch fixes an issue in the Rockchip PCIe PHY driver where, after a warm restart of the rockchip_pcie_phy module, PCIe lanes other than lane 0 could remain stuck in the PHY_LANE_IDLE_OFF state. This resulted in the PCIe link being restricted to x1 mode, even in configurations designed to use multiple lanes.
++
++Signed-off-by: Valmantas Paliksa <walmis@gmail.com>
++---
++ drivers/phy/rockchip/phy-rockchip-pcie.c | 12 ++++++------
++ 1 file changed, 6 insertions(+), 6 deletions(-)
++
++diff --git a/drivers/phy/rockchip/phy-rockchip-pcie.c b/drivers/phy/rockchip/phy-rockchip-pcie.c
++index 8234b83fdd88..240cb27a0e9e 100644
++--- a/drivers/phy/rockchip/phy-rockchip-pcie.c
+++++ b/drivers/phy/rockchip/phy-rockchip-pcie.c
++@@ -166,6 +166,12 @@
++ 	unsigned long timeout;
++ 
++ 	mutex_lock(&rk_phy->pcie_mutex);
+++	
+++	regmap_write(rk_phy->reg_base,
+++		     rk_phy->phy_data->pcie_laneoff,
+++		     HIWORD_UPDATE(!PHY_LANE_IDLE_OFF,
+++				   PHY_LANE_IDLE_MASK,
+++				   PHY_LANE_IDLE_A_SHIFT + inst->index));
++ 
++ 	if (rk_phy->pwr_cnt++)
++ 		goto err_out;
++@@ -181,12 +187,6 @@
++ 				   PHY_CFG_ADDR_MASK,
++ 				   PHY_CFG_ADDR_SHIFT));
++ 
++-	regmap_write(rk_phy->reg_base,
++-		     rk_phy->phy_data->pcie_laneoff,
++-		     HIWORD_UPDATE(!PHY_LANE_IDLE_OFF,
++-				   PHY_LANE_IDLE_MASK,
++-				   PHY_LANE_IDLE_A_SHIFT + inst->index));
++-
++ 	/*
++ 	 * No documented timeout value for phy operation below,
++ 	 * so we make it large enough here. And we use loop-break
+-- 
+2.47.1
+
-- 
2.47.1

